### PR TITLE
Upgrade to 1.14, add new tags for use with custom resource packs

### PR DESCRIPTION
**Pull request by Pangamma**
- Upgrade to 1.14
- Add new tags for custom resource packs. (New additions will ultimately affect: damaged, damage, and custom_model_data when viewed through predicates for resource packs. To affect each, use DAMAGE, DAMAGED/UNBREAKABLE/UNBREAKING, CUSTOM-MODEL-DATA. 
- Better data clean-up when plugin unloads.
image
image

### CODE REVIEW

File: Plugin/src/main/java/com/gmail/filoghost/chestcommands/ChestCommands.java [hunk 1]
```diff
@@ -146,6 +147,19 @@ public void onUpdateFound(String newVersion) {
 	@Override
 	public void onDisable() {
 		closeAllMenus();
+		HandlerList.unregisterAll(this);
```
Comment (Pangamma) [inline Plugin/src/main/java/com/gmail/filoghost/chestcommands/ChestCommands.java]: Unregisters listeners

**Comment by Pangamma** [comment]
image

File: Plugin/src/main/java/com/gmail/filoghost/chestcommands/ChestCommands.java (hunk 1 above)
Comment (sgdc3) [inline Plugin/src/main/java/com/gmail/filoghost/chestcommands/ChestCommands.java]: Bukkit already does this

File: Plugin/src/main/java/com/gmail/filoghost/chestcommands/ChestCommands.java [hunk 2]
```diff
@@ -146,6 +147,19 @@ public void onUpdateFound(String newVersion) {
 	@Override
 	public void onDisable() {
 		closeAllMenus();
+		HandlerList.unregisterAll(this);
+		Bukkit.getScheduler().cancelTasks(this);
+		getServer().getServicesManager().unregisterAll(this);
+		this.getServer().getMessenger().unregisterIncomingPluginChannel(this);
+		this.getServer().getMessenger().unregisterOutgoingPluginChannel(this);
+		instance = null;
```
Comment (sgdc3) [inline Plugin/src/main/java/com/gmail/filoghost/chestcommands/ChestCommands.java]: It makes no sense to nullify the instances

File: Plugin/src/main/resources/plugin.yml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 name: ChestCommands
 main: com.gmail.filoghost.chestcommands.ChestCommands
 version: ${project.version}
-api-version: 1.13
+api-version: 1.14
```
Comment (sgdc3) [inline Plugin/src/main/resources/plugin.yml]: Nope, this breaks 1.13 compatibility

File: pom.xml [hunk 4]
```diff
@@ -22,7 +22,7 @@
 		<maven.compiler.source>1.6</maven.compiler.source>
 		<maven.compiler.target>1.6</maven.compiler.target>
 		<maven.test.skip>true</maven.test.skip>
-		<spigot-api.version>1.8-R0.1-SNAPSHOT</spigot-api.version>
+		<spigot-api.version>1.14-R0.1-SNAPSHOT</spigot-api.version>
```
Comment (sgdc3) [inline pom.xml]: Please keep this to the minimum supported api version

File: Plugin/pom.xml [hunk 5]
```diff
@@ -64,6 +64,7 @@
 	</dependencies>
 
 	<build>
+        <finalName>ChestCommands</finalName>
```
Comment (sgdc3) [inline Plugin/pom.xml]: Identation is wrong here, filoghost used tabs, not spaces.

**Comment by sgdc3** [comment]
This PR should be merged after mine, as it would be easier to fix conflicts in this one, as it changes less files.

File: Plugin/src/main/java/com/gmail/filoghost/chestcommands/ChestCommands.java (hunk 2 above)
Comment (sgdc3) [reply Plugin/src/main/java/com/gmail/filoghost/chestcommands/ChestCommands.java]: Services and messaging channels are already unregistered by bukkit

**Comment by filoghost** [comment]
Hi, thanks for the submission. In addition to @sgdc3 suggestions, I ask to only add tags that are useful for the end user **and** are simpler than just using NBT-DATA (for example, enchantments).

What would UNBREAKABLE be used for? The menu cannot be modified by a player. CUSTOM-MODEL-DATA is an edge case, few servers use custom textures for menus. Same for DAMAGE, I don't see many use cases. All of these tags should be used through NBT-DATA, unless they're frequently used.

**Comment by Pangamma** [comment]
@filoghost @sgdc3  I tried for days to get the NBT-TAG thing to work. Nothing I tried was working. Perhaps you could provide a working example? Seriously. I've tried. Nothing I've tried has worked. 

I did make a new PR though. It has the original requested changes in it. Could we please merge it in, or else provide some super specific (and tested) example usage values for the NBT-TAG thing for the use-case I'm trying to handle right now? 
https://github.com/filoghost/ChestCommands/pull/17

In response to the, "Why would we need these tags?" question, the tags match up with these predicate values within resource packs. Some odd values are needed to make sure the packs show up correctly. 
https://minecraft.gamepedia.com/Model#Item%20models

**Comment by Pangamma** [comment]
[ModelPackResourcePack.zip](https://github.com/filoghost/ChestCommands/files/3160453/ModelPackResourcePack.zip)
[chest-commands.zip](https://github.com/filoghost/ChestCommands/files/3160455/chest-commands.zip)

Here. Give it a try. Should work on a 1.14 server. There is a resource pack here, and the chest command configuration files for them.

**Comment by filoghost** [comment]
I don't know what you tried, but I got it working on the first try for CustomModelData:
```custom:
  ID: diamond_sword
  NBT-DATA: '{CustomModelData:1002}'
  POSITION-X: 1
  POSITION-Y: 1
```

I have not been able to try the old "Damage" method, because I couldn't get it to work even in vanilla. Please provide me a /give command for it that works in 1.14.

**Comment by Pangamma** [comment]
And now we will have documentation for future Googlers. Awesome.

**Pull request closed without merging**